File: JavaScript/0144. Binary Tree Preorder Traversal.js
// Given a binary tree, return the preorder traversal of its nodes' values.
//
// Example:
//
//   Input: [1,null,2,3]
//     1
//      \
//       2
//      /
//     3
//
// Output: [1,2,3]
//
// Follow up: Recursive solution is trivial, could you do it iteratively?

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

/** 1) Recursion */
const preorderTraversal1 = (root) => {
  const res = [];

  const go = (node) => {
    if (node == null) return;
    res.push(node.val);
    go(node.left);
    go(node.right);
  };

  go(root);
  return res;
};

/** 2) Iteration using stack */
const preorderTraversal = (root) => {
  const st = [];
  const res = [];

  while (root || st.length) {
    // print & drill left
    while (root) {
      res.push(root.val);
      st.push(root);
      root = root.left;
    }

    // go to right child
    root = st.pop();
    root = root.right;
  }

  return res;
};

File: Python/0144. Binary Tree Preorder Traversal.py
# Given the root of a binary tree, return the preorder traversal of its nodes' values.
#
# Example 1:
#
# Input: root = [1,null,2,3]
# Output: [1,2,3]
#
# Example 2:
#
# Input: root = []
# Output: []
#
# Example 3:
#
# Input: root = [1]
# Output: [1]
#
# Constraints:
#
# The number of nodes in the tree is in the range [0, 100].
# -100 <= Node.val <= 100


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 1) Recursion
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []
        return (
            [root.val]
            + self.preorderTraversal(root.left)
            + self.preorderTraversal(root.right)
        )


# 2) Iteration using stack
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        res = []
        st = [root]

        while st:
            node = st.pop()
            res.append(node.val)
            if node.right:
                st.append(node.right)
            if node.left:
                st.append(node.left)
        return res
